packages/loopring_v3/contracts/amm/libamm/AmmJoinProcess.sol

```diff
@@ -0,0 +1,211 @@
+                    refundAmount = transfer.amount.sub(amount);
+                }
+
+                if (refundAmount > 0) { // Process the outbound transfer
+                    transfer = ctx._block.readTransfer(ctx.txIdx++);
+                    ctx.numTransactionsConsumed++;
+
+                    // We do not check these fields: toAccountID
+                    require(
+                        transfer.fromAccountID == S.accountID &&
+                        transfer.from == address(this) &&
+                        transfer.to == join.owner &&
+                        transfer.tokenID == ctx.tokens[i].tokenID &&
+                        transfer.amount.isAlmostEqual(refundAmount) &&
+                        transfer.feeTokenID == 0 &&
+                        transfer.fee == 0 &&
+                        transfer.storageID == 0, // Question (brecht):should we check this?
+                        "INVALID_OUTBOUND_TX_DATA"
+                    );
+
+                   // Now approve this transfer
+                    transfer.validUntil = 0xffffffff;
+                    txHash = TransferTransaction.hashTx(ctx.exchangeDomainSeparator, transfer);
+                    ctx.exchange.approveTransaction(address(this), txHash);
+                    amount = amount.sub(transfer.amount);
+                }
+
+                ctx.ammActualL2Balances[i] = ctx.ammActualL2Balances[i].add(amount);
+
+            } else {
+                // Make the amount unavailable for withdrawing
+                address token = ctx.tokens[i].addr;
+                S.lockedBalance[token][join.owner] = S.lockedBalance[token][join.owner].sub(amount);
+            }
+
+            ctx.ammExpectedL2Balances[i] = ctx.ammExpectedL2Balances[i].add(amount);
+        }
+
+        S.mint(join.owner, poolAmountOut);
+    }
+
+    function _calculateJoinAmounts(
+        AmmData.Context  memory ctx,
+        AmmData.PoolJoin memory join
+        )
+        private
+        view
+        returns(
+            bool            slippageRequirementMet,
+            uint            poolAmountOut,
+            uint96[] memory amounts
+        )
+    {
+        // Check if we can still use this join
+        amounts = new uint96[](ctx.size);
+
+        if (block.timestamp > join.validUntil) {
+            return (false, 0, amounts);
+        }
+
+        if (ctx.poolTokenTotalSupply == 0) {
+            return(true, ctx.poolTokenInitialSupply, join.maxAmountsIn);
+        }
+
+        // Calculate the amount of liquidity tokens that should be minted
+        for (uint i = 0; i < ctx.size; i++) {
+            if (ctx.ammExpectedL2Balances[i] > 0) {
+                uint amountOut = uint(join.maxAmountsIn[i])
+                    .mul(ctx.poolTokenTotalSupply) / uint(ctx.ammExpectedL2Balances[i]);
+
+                if (poolAmountOut == 0 || amountOut < poolAmountOut) {
+                    poolAmountOut = amountOut;
+                }
```

Comment:
Using `poolAmountOut == 0` like this is not necessarily correct here. `maxAmountsIn[0]` could be `0` which would correctly set `poolAmountOut` to `0`, but then it could still be incorrectly increased by following `maxAmountsIn`s.

This example is of course not very realistic because why would a user set `maxAmountsIn == 0`, but this is still a valid input and with rounding errors and tokens with low decimals I think there are cases this could end up being wrong.